### PR TITLE
server: sotw dedupe and linearization

### PR DESCRIPTION
**Pull request by alecholmez**
Signed-off-by: Alec Holmes <alecholmez@me.com>

I closed my previous sotw dedupe PR in favor of this one since we merged in the linearization API code which this now adheres too. Still need to benchmark

### CODE REVIEW

**Comment by dontlaugh** [comment]
@alecholmez you want to do benchmarks for this, right

**Comment by alecholmez** [comment]
@dontlaugh yup, we'd like to see how this compares with the extra go routine overhead to the current system

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:60]: Do we need to track it here in addition to cases?

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:48]: len(newCases) might seem less magical

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:35]: I would just say rebuilds the list of dynamic channels, I think it could go both up and down

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:21]: This comment seems out of date?

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:24]: Don't we want this to be properly initialized? I assume we'll call RecompueWatches before this ever get read, so maybe just call it directly in newWatches or avoid this logic since it doesn't matter?

File: pkg/server/sotw/v3/watches.go
Comment (snowp) [inline pkg/server/sotw/v3/watches.go:67]: When do we use this?

File: pkg/server/v3/server_test.go
Comment (snowp) [inline pkg/server/v3/server_test.go:374]: Seems like it would be nicer to have these be
```
assert.NotNil(t, out)
assert.NoError(t, err)
```
?

File: pkg/server/v3/server_test.go
Comment (snowp) [inline pkg/server/v3/server_test.go:438]: assert.Equal

File: pkg/server/v3/server_test.go
Comment (snowp) [inline pkg/server/v3/server_test.go:593]: Equal

File: pkg/server/v3/server_test.go
Comment (snowp) [inline pkg/server/v3/server_test.go:611]: Equal

**Review by snowp — COMMENTED** [review]
Thanks this looks good for the most part, just a few comments

**Review by snowp — DISMISSED** [review]
Seems like CI is red?

File: pkg/server/sotw/v3/server.go
Comment (alecholmez) [reply pkg/server/sotw/v3/server.go:202]: Ah yea that is quite a bit more readable I'll move to that.

File: pkg/server/sotw/v3/watches.go
Comment (alecholmez) [reply pkg/server/sotw/v3/watches.go:51]: Do you imagine that being too expensive at scale? I don't really ever imagine that slice becoming too large to the point where it's particularly an issue recalculating it, but when 10k+ streams are open and watches are being opened and canceled a lot that might be a problem. Istio has an issue like this right now where they see massive traffic and data movement (lots of config updates) on startup but once the mesh stabilizes it doesn't really become an expensive operation anymore. Unless things come up and down of course. What are your thoughts on that? I'd imagine this recompute would get called a lot in that scenario. I'm more inclined to deal with a bit more complexity on the code side for a little edge in performance

File: pkg/server/sotw/v3/watches.go
Comment (alecholmez) [reply pkg/server/sotw/v3/watches.go:35]: I think a `watch` in this particular server code sense is more of an internal data behavior. In the previous code, what the `watch` and `watches` struct replace are overly repeated bits that are for moving data around (more or less cache interaction).

If you do have suggestions for how we can streamline that I'd love to hear them. I mostly just refactored that old code to these new structs simply to reduce repetitive lines with similar/same functionality (nothing too special on that front :joy:).

Comment (jpeach) [reply pkg/server/sotw/v3/watches.go:35]: Yeh, I'd suggest renaming `RecomputeWatches()` to `recompute()`, and `Cancel()` to `cancel()` on both `watches` and `watch`. To me, that makes it really clear that this is all internal implementation details,

File: pkg/server/sotw/v3/watches.go
Comment (jpeach) [reply pkg/server/sotw/v3/watches.go:51]: I tried this out and IMHO it seems cleaner ([gist](https://gist.github.com/jpeach/128b5199a564928d396545f92f7fc3a8)).

There's one watch per typeURL (so maybe a dozen would be a lot?), so it doesn't seem likely that this sort of micro-optimization would have a big impact.

File: pkg/server/sotw/v3/watches.go
Comment (alecholmez) [reply pkg/server/sotw/v3/watches.go:35]: That makes sense. I think as the code stands the `cancel()` method on watch will cause a name collision with the field `cancel func()`. Maybe rename those to close instead?

File: pkg/server/sotw/v3/watches.go
Comment (alecholmez) [reply pkg/server/sotw/v3/watches.go:51]: Awesome thanks for that diff :) I just made the changes and I'll push

**Comment by alecholmez** [comment]
@jpeach can you give this another pass when you get some free time?

File: pkg/server/sotw/v3/watches.go
Comment (jpeach) [inline pkg/server/sotw/v3/watches.go:16]:
```suggestion
	// cases is a dynamic select case which matches the watches.
```

**Review by jpeach — DISMISSED** [review]
:+1:

**Review by jpeach — APPROVED** [review]

**Comment by alecholmez** [comment]
@snowp want to give the final stamp on this one?

**Comment by rueian** [comment]
Hi @alecholmez, @jpeach,

I based on this PR to solve another deadlock in sotw server (#530, #531). Please take a look.

**Comment by alecholmez** [comment]
@jpeach is this good to merge? @rueian once this is merged can you rebase your branch off main instead?

**Comment by jpeach** [comment]
> @jpeach is this good to merge?

Yup!

**Review by jpeach — APPROVED** [review]

**Pull request merged**
